lib/pbench/server/api/resources/metadata_api.py

```diff
@@ -0,0 +1,355 @@
+        This requires a Pbench auth token in the header field
+
+        Required headers include
+            Authorization:   Bearer <Pbench authentication token (user received upon login)>
+
+        :return: JSON Payload
+            response_object = {
+                    "message": "success"
+                    "data": {
+                        "sessions": [
+                            {"id": "metadata_id",
+                            "value": "client text blob",
+                            "key": "key string",
+                            "created": "datetime string",
+                            "updated": "datetime string", },
+                            {}]
+                        }
+                }
+        """
+        if not key:
+            self.logger.warning("Metadata key not provided during metadata query")
+            abort(400, message="Missing metadata key in the query url")
+
+        current_user = self.auth.token_auth.current_user()
+        if current_user:
+            current_user_id = current_user.id
+        else:
+            current_user_id = None
+        print(current_user_id)
+        try:
+            # Fetch the metadata sessions
+            # If the key is favorite, we return only favorite sessions,
+            # else we return all the saved and favorite sessions
+            if key.upper() == "FAVORITE":
+                metadata_sessions = Metadata.query(
+                    user_id=current_user_id, key=MetadataKeys[key.upper()].value
+                )
+            else:
+                metadata_sessions = Metadata.query(user_id=current_user_id)
+            data = [session.get_json() for session in metadata_sessions]
+        except Exception:
+            self.logger.exception(
+                "Exception occurred while querying the Metadata model"
+            )
+            abort(500, message="INTERNAL ERROR")
+
+        response_object = {
+            "message": "success",
+            "data": {"sessions": data},
+        }
+        return make_response(jsonify(response_object), 200)
+
+
+class QueryMetadata(Resource):
+    """
+    Abstracted pbench API for querying a single user metadata session
+    """
+
+    def __init__(self, config, logger):
+        self.server_config = config
+        self.logger = logger
+
+    def verify_metadata(self, session, action):
+        current_user_id = Auth.token_auth.current_user().id
+        metadata_user_id = session.user_id
+        if current_user_id != metadata_user_id:
+            self.logger.warning(
+                "Metadata {}: Logged in user_id {} is different than the one provided in the URI {}",
+                action,
+                current_user_id,
+                metadata_user_id,
+            )
+            abort(403, message="Not authorized to perform the specified action")
+
+    @Auth.token_auth.login_required()
+    def get(self, id=None):
```

Comment:
If template values in URIs are really optional, another way to handle this might be with a single `/metadata/<key>/<id>` mapping to a single resource object; if both values are missing, you return all metadata for the user; if there's only <key>, return all values for that key; if only <id>, return the specific ID. (And `/metadata/favorite/2` is fine as long as metadata row 2 is a "favorite".)

Reply:
Template values in URIs are _not_ really optional, they are optional in a sense that it can take any value but the value need to be there. 

Having an endpoint `/metadata/<key>/<id>` to get certain metadata object, dashboard need to know that which id belong to which key. I am not aware if there is anything on dashboard to save this knowledge to effectively query the metadata object. I think the better option would be `/metadata/<id>` to get the object and let the dashboard decide if this is the right one. The dashboard can query the list of all the favorites as `/metadata/favorite` and then get details of specific metadata object by querying on `/metadata/<id>`

Reply:
I'm trying to picture how this works from a user and client perspective.  I'm presuming that the client presents the user with a list of favorites to choose from...how does it construct that list?  Well, presumably it does a `GET` on `/metadata/favorites` which returns all of the metadata (with their corresponding ID's) for that user.  And, then, somehow, the user selects one, for which the client now has the ID from the previous request, and the client issues a `GET` on `/metadata/favorites/<id>`.

I'm not really coming up with a scenario where the client has an ID but does _not_ know what key it goes with, since the client used the key when it received the ID value.